src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_validators.py

```diff
@@ -52,22 +65,101 @@ def validate_client_parameters(namespace):
         else:
             raise ValueError("Storage account '{}' not found.".format(n.account_name))
 
-def get_content_setting_validator(settings_class):
+def validate_source_uri(namespace):
+    usage_string = 'invalid usage: supply only one of the following argument sets:' + \
+                   '\n\t   --source-uri' + \
+                   '\n\tOR --source-container --source-blob [--source-snapshot] [--source-sas]' + \
+                   '\n\tOR --source-share --source-path [--source-sas]'
+    ns = vars(namespace)
+    validate_client_parameters(namespace) # must run first to resolve storage account
+    storage_acc = ns.get('account_name', None) or os.environ.get('AZURE_STORAGE_ACCOUNT')
+    uri = ns.get('copy_source', None)
+    container = ns.pop('source_container', None)
+    blob = ns.pop('source_blob', None)
+    sas = ns.pop('source_sas', None)
+    snapshot = ns.pop('source_snapshot', None)
+    share = ns.pop('source_share', None)
+    path = ns.pop('source_path', None)
+    if uri:
+        if any([container, blob, sas, snapshot, share, path]):
+            raise ValueError(usage_string)
+        else:
+            # simplest scenario--no further processing necessary
+            return
+
+    valid_blob_source = container and blob and not share and not path
+    valid_file_source = share and path and not container and not blob and not snapshot
+
+    if valid_blob_source:
+        blob_uri = 'https://{}.blob.core.windows.net/{}/{}{}'.format(
+            storage_acc, container, blob, '?' if snapshot or sas else '')
+        if snapshot:
+            blob_uri = '{}snapshot={}'.format(blob_uri, snapshot)
+        if sas:
+            blob_uri = '{}{}{}'.format(blob_uri, '&' if snapshot else '', sas)
+        namespace.copy_source = blob_uri
+    elif valid_file_source:
+        file_uri = 'https://{}.file.core.windows.net/{}/{}'.format(
+            storage_acc, share, path)
+        if sas:
+            file_uri = '{}?{}'.format(file_uri, sas)
+        namespace.copy_source = file_uri
+    else:
+        raise ValueError(usage_string)
+
+def get_content_setting_validator(settings_class, update):
+    def _class_name(class_type):
+        return class_type.__module__ + "." + class_type.__class__.__name__
+
     def validator(namespace):
-        namespace.content_settings = settings_class(
-            content_type=namespace.content_type,
-            content_disposition=namespace.content_disposition,
-            content_encoding=namespace.content_encoding,
-            content_language=namespace.content_language,
-            content_md5=namespace.content_md5,
-            cache_control=namespace.content_cache_control
+        # must run certain validators first for an update
+        if update:
+            validate_client_parameters(namespace)
+        if update and _class_name(settings_class) == _class_name(FileContentSettings):
+            get_file_path_validator()(namespace)
```

Comment:
No. 'get_file_path_validator' returns a validator and then (namespace) runs it on the namespace.

---
In reply to: [77395762](https://github.com/Azure/azure-cli/pull/804#discussion_r77395762) [](ancestors = 77395762)